extensions/amp-sidebar/1.0/sidebar.js

```diff
@@ -0,0 +1,216 @@
+import {ContainWrapper} from '../../../src/preact/component';
+import {forwardRef} from '../../../src/preact/compat';
+import {setStyle} from '../../../src/style';
+import {
+  useCallback,
+  useEffect,
+  useImperativeHandle,
+  useRef,
+  useState,
+} from '../../../src/preact';
+import {useStyles} from './sidebar.jss';
+
+const ANIMATION_DURATION = 350;
+const ANIMATION_PRESETS = {
+  fadeIn: [{opacity: 0}, {opacity: 1}],
+  slideInLeft: [{transform: 'translateX(-100%)'}, {transform: 'translateX(0)'}],
+  slideInRight: [{transform: 'translateX(100%)'}, {transform: 'translateX(0)'}],
+};
+
+const ANIMATION_EASE_IN = 'cubic-bezier(0,0,.21,1)';
+
+/**
+ * @param {T} current
+ * @return {{current: T}}
+ * @template T
+ */
+function useValueRef(current) {
+  const valueRef = useRef(null);
+  valueRef.current = current;
+  return valueRef;
+}
+
+/**
+ * @param {!AccordionDef.Props} props //TODO: will update
+ * @param ref
+ * @return {PreactDef.Renderable}
+ */
+function SidebarWithRef(
+  {
+    as: Comp = 'div',
+    side = 'left',
+    onBeforeOpen,
+    onAfterClose,
+    children,
+    ...rest
+  },
+  ref
+) {
+  // There are two phases to open and close.
+  // To open, we mount and render the contents (invisible), then animate the display (visible).
+  // To close, it's the reverse.
+  // `mounted` mounts the component. `opened` plays the animation.
+  const initialOpen = false;
+  const [mounted, setMounted] = useState(initialOpen);
+  const [opened, setOpened] = useState(initialOpen);
+  const classes = useStyles();
+  const sidebarRef = useRef();
+  const maskRef = useRef();
+
+  // We are using refs here to refer to common strings, objects, and functions used.
+  // This is because they are needed within `useEffect` calls below (but are not depended for triggering)
+  // We use `useValueRef` for props that might change (user-controlled)
+  const onBeforeOpenRef = useValueRef(onBeforeOpen);
+  const onAfterCloseRef = useValueRef(onAfterClose);
+
+  const open = useCallback(() => {
+    if (onBeforeOpenRef.current) {
+      onBeforeOpenRef.current();
+    }
+    setMounted(true);
+    setOpened(true);
+  }, [onBeforeOpenRef]);
+  const close = useCallback(() => setOpened(false), []);
+  const toggle = useCallback(() => (opened ? close() : open()), [
+    opened,
+    open,
+    close,
+  ]);
+
+  useImperativeHandle(
+    ref,
+    () => ({
+      open,
+      close,
+      toggle,
+    }),
+    [open, close, toggle]
+  );
+
+  useEffect(() => {
+    const sidebarElement = sidebarRef.current;
+    const maskElement = maskRef.current;
+    if (sidebarElement == undefined || maskElement == undefined) {
+      return;
+    }
+
+    let sidebarAnimation;
+    let maskAnimation;
+    // "Make Visible" Animation
+    if (opened) {
+      const postVisibleAnim = () => {
+        setStyle(sidebarElement, 'transform', 'translateX(0)');
+        setStyle(maskElement, 'opacity', 1);
+        sidebarElement./*REVIEW*/ focus();
```

Comment:
Does this do anything at all? Is the intention to focus the first focusable element in the sidebar? That'd be a good idea, of course, but I'm not sure this does that.

Reply:
I will remove for now and address in a future PR.

In the previous sidebar, the focus is shifted to an invisible close button.

Reply:
That's actually a good idea. But if it's coming in the follow up PRs, feel free to defer.

Reply:
Thanks, added to the tracker item: https://github.com/ampproject/amphtml/issues/31366